internal/util/recoverable.go

```diff
@@ -0,0 +1,113 @@
+package util
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"runtime/debug"
+	"sync"
+	"time"
+)
+
+var (
+	ErrServiceStopped = fmt.Errorf("service stopped")
+)
+
+type Doable interface {
+	Do() error
+	Stop()
+}
+
+func NewRecoverableService(svc Doable, logger *log.Logger) *RecoverableService {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &RecoverableService{
+		service: svc,
+		stopped: make(chan stoppedService, 1),
+		log:     logger,
+		ctx:     ctx,
+		cancel:  cancel,
+	}
+}
+
+type RecoverableService struct {
+	mu      sync.Mutex
+	running bool
+	service Doable
+	stopped chan stoppedService
+	log     *log.Logger
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+func (m *RecoverableService) Start(_ context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.running {
+		return nil
+	}
+
+	go m.serviceStart()
+	m.running = true
+	return nil
+}
+
+func (m *RecoverableService) Stop() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.running {
+		return
+	}
+
+	m.service.Stop()
+	m.cancel()
+	m.running = false
+}
+
+func (m *RecoverableService) serviceStart() {
+	for {
+		select {
+		case st := <-m.stopped:
+			// restart the service
+			if st.Err != nil && errors.Is(st.Err, ErrServiceStopped) {
+				<-time.After(10 * time.Second)
+				m.run()
+			}
+		case <-m.ctx.Done():
+			return
+		}
+	}
+}
+
+func (m *RecoverableService) run() {
+	go func(s Doable, l *log.Logger, chStop chan stoppedService, ctx context.Context) {
+		defer func() {
+			if err := recover(); err != nil {
```

Comment:
Just nitpicking here, but could we rename this `err` to `r` or `rec`, since it's not really an error?

Reply:
According to the docs, it is an error. The return from `recover()` is the error passed to `panic()`. At least that's my understanding...

```
The recover built-in function allows a program to manage behavior of a panicking goroutine. Executing a call to recover inside a deferred function (but not any function called by it) stops the panicking sequence by restoring normal execution and retrieves the error value passed to the call of panic. If recover is called outside the deferred function it will not stop a panicking sequence. In this case, or when the goroutine is not panicking, or if the argument supplied to panic was nil, recover returns nil. Thus the return value from recover reports whether the goroutine is panicking.
```

Reply:
It's weird that it returns an `any` type even though its an error. Maybe an `any` type can be passed to `panic()`??? Never looked into that. ¯\_(ツ)_/¯